packages/multicast_dns/lib/src/packet.dart

```diff
@@ -0,0 +1,329 @@
+// parts and the number of bytes consumed.
+//
+// If decoding fails (e.g. due to an invalid packet) `null` is returned.
+_FQDNReadResult _readFQDN(Uint8List data, ByteData bd, int offset, int length) {
+  void checkLength(int required) {
+    if (length < required) {
+      throw MDnsDecodeException(required);
+    }
+  }
+
+  final List<String> parts = <String>[];
+  final int prevOffset = offset;
+  while (true) {
+    // At least one byte is required.
+    checkLength(offset + 1);
+
+    // Check for compressed.
+    if (data[offset] & 0xc0 == 0xc0) {
+      // At least two bytes are required for a compressed FQDN.
+      checkLength(offset + 2);
+
+      // A compressed FQDN has a new offset in the lower 14 bits.
+      final _FQDNReadResult result =
+          _readFQDN(data, bd, bd.getUint16(offset) & ~0xc000, length);
+      parts.addAll(result.fqdnParts);
+      offset += 2;
+      break;
+    } else {
+      // A normal FQDN part has a length and a UTF-8 encoded name
+      // part. If the length is 0 this is the end of the FQDN.
+      final int partLength = data[offset];
+      offset++;
+      if (partLength > 0) {
+        checkLength(offset + partLength);
+        final Uint8List partBytes =
+            Uint8List.view(data.buffer, offset, partLength);
+        offset += partLength;
+        parts.add(utf8.decode(partBytes));
+      } else {
+        break;
+      }
+    }
+  }
+  return _FQDNReadResult(parts, offset - prevOffset);
+}
+
+/// Decode a mDNS package.
+///
+/// If decoding fails (e.g. due to an invalid packet) `null` is returned.
+///
+/// See https://tools.ietf.org/html/rfc1035 for the format.
+List<ResourceRecord> decodeMDnsResponse(List<int> packet) {
+  final int length = packet.length;
+  if (length < _kHeaderSize) {
+    return null;
+  }
+
+  final Uint8List data =
+      packet is Uint8List ? packet : Uint8List.fromList(packet);
+  final ByteData bd = ByteData.view(data.buffer);
+  // Number of answers.
+  final int ancount = bd.getUint16(_kAncountOffset);
+  if (ancount == 0) {
+    return null;
+  }
+
+  // Number of resource records.
+  final int arcount = bd.getUint16(_kArcountOffset);
+  int offset = _kHeaderSize;
+
+  void checkLength(int required) {
+    if (length < required) {
+      throw MDnsDecodeException(required);
+    }
+  }
+
+  ResourceRecord readResourceRecord() {
+    // First read the FQDN.
+    final _FQDNReadResult result = _readFQDN(data, bd, offset, length);
+    final String fqdn = result.fqdn;
+    offset += result.bytesRead;
+    checkLength(offset + 2);
+    final int type = bd.getUint16(offset);
+    offset += 2;
+    // The first bit of the rrclass field is set to indicate that the answer is
+    // unique and the querier should flush the cached answer for this name
+    // (RFC 6762, Sec. 10.2). We ignore it for now since we don't cache answers.
+    checkLength(offset + 2);
+    final int cls = bd.getUint16(offset) & 0x7fff;
```

Comment:
avoid abbreviations. i've no idea, glancing at the code, what `bd` and `cls` are.